### PR TITLE
Foundation for UDMIS work

### PR DESCRIPTION
**Pull request by grafnu**
Very little actual functionality -- `StateHandler` is the only class that's "interesting" in terms of actual UDMIS behavior. However, the primary goal of this PR is more foundational work in terms of setting up a solid base to properly handle basic capabilities that can be used going forward. Specific items to think about:

1. Unit testing and design for testability
2. Logging mechanism and abstraction
3. Message sending/handling abstractions
4. Containerization design (not containerized yet, but just structured around the concept of a `pod`)

Specifically, the `StateHandler` class is meant to be as "elegant" as possible and focus on the business logic (which is very simple in this case), while everything else is "supporting" and wouldn't be expected to change very much as additional business logic is included.

This isn't really about "UDMIS" in terms of functionality there -- there's very little of it -- as that is still pending the design doc proper.

### CODE REVIEW

File: udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * Tests for the StateHandler class, used by UDMIS to process device state updates.
+ */
+public class StateHandlerTest extends MessageTestBase {
+
+  public static final String INVALID_MESSAGE = "invalid message";
+  private StateHandler stateHandler;
+
+  @Test
```
Comment (johnrandolph) [inline udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java:36]: All helper functions first then tests?

File: udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  public void singleExpansion() throws InterruptedException {
```
Comment (johnrandolph) [inline udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java:37]: Add comments for all test methods describing what they do?

File: udmis/bin/build [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  fi
+done
+
+ROOT=$(dirname $0)/..
+BASE=..
+cd $ROOT
+
+jarfile=build/libs/udmis-1.0-SNAPSHOT-all.jar
+
+newest=$(ls -rt `find src/ $BASE/gencode/java -type f` | tail -n 1)
```
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/bin/build:21]: it's newest, the -r reverses the order. I suppose it would be the same as "ls -t | head -n 1" -- but I always did it the other way since I usually use the same at the command line (easier to see the last line output than the first)

File: udmis/src/main/java/com/google/bos/udmi/service/pod/ComponentBase.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.google.bos.udmi.service.pod;
```
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/src/main/java/com/google/bos/udmi/service/pod/ComponentBase.java:1]: https://kubernetes.io/docs/concepts/workloads/pods/ -- something like that -- just a clump of stuff in terms of managing deployment, but it's not the functional/business logic itself.

File: udmis/src/main/java/com/google/bos/udmi/service/core/StateHandler.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        if (STATE_SUB_FOLDERS.contains(field.getName())) {
+          ifNotNull(field.get(message), x -> pipe.publish(x));
+        }
+      } catch (Exception e) {
+        throw new RuntimeException("While extracting field " + field.getName(), e);
+      }
+    });
+  }
+
+  public static StateHandler forConfig(MessageConfiguration configuration) {
```
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/src/main/java/com/google/bos/udmi/service/core/StateHandler.java:79]: Honestly, I never pay attention to the ordering of methods. It's one of those if-the-IDE-does-it-or-the-linter-catches-it then I'll fix it kinds of things. I'd say if there was a linter option you could find that enforced it then I'd be for it, but otherwise I don't see how it practically is enforceable.

File: udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java (hunk 1 above)
Comment (grafnu) [reply udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java:36]: As per other comment -- not something I ever pay attention to, so unless there's something automatic about detecting it I doubt it'll be a consistent convention. I can try to rearrange some, but it just seems like unnecessary overhead to me.

File: udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java (hunk 2 above)
Comment (grafnu) [reply udmis/src/test/java/com/google/bos/udmi/service/core/StateHandlerTest.java:37]: Can do, will update and flag for re-review.

File: udmis/src/main/java/com/google/bos/udmi/service/core/StateHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/src/main/java/com/google/bos/udmi/service/core/StateHandler.java]: Ok, I've done my best to put things in a logical order. Honestly it just doesn't seem like something by brain works with (I would stare at the methods and nothing seemed like a good linear order)... I think I generally just don't think of code in a linear way so it kinda trips me up!  PTAL and let me know if you think I should take another pass.

**Review by johnrandolph — COMMENTED** [review]
This IntelliJ code cleanup mode is a great improvement for UDMI overall.

Just checking, in a broader sense, should we be using some off the shelf message queue rather than writing our own? The PR would be smaller, IMHO, if it basically centered around the Udmi Pod, State Handler, and the definition for what the messages look like in the queue.

File: udmis/src/test/java/com/google/bos/udmi/service/messaging/MessageTestBase.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    MessageBase messagePipe = getTestMessagePipeCore(reversed);
+    if (!reversed && !messagePipe.isActive()) {
+      messagePipe.registerHandlers(messageHandlers);
+      messagePipe.activate();
+    }
+    return messagePipe;
+  }
+
+  private <T> void messageHandler(T message) {
+    synchronized (LocalMessagePipeTest.class) {
```
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/src/test/java/com/google/bos/udmi/service/messaging/MessageTestBase.java]: No it should not -- updated/fixed.

**Comment by grafnu** [comment]
> Just checking, in a broader sense, should we be using some off the shelf message queue rather than writing our own? 

If you have a viable alternative please let me know!  Yes, I agree using a off-the-shelf solution would be better, but it's not really a message queue that this provides, it's an _abstraction_ around multiple different message queues (mqtt, pubsub, etc...) -- so that the use points don't get tied into the underlying queue implementations themselves. The closest I can find is [JMS](https://docs.oracle.com/javaee/6/api/javax/jms/package-summary.html), which is in theory an interface layer but seems to pull in way more complexity than we actually need.  It's the usual problem I find with these abstraction libraries out there (sl4j) in that they start off doing something simple (logging interface) but then get really complicated (e.g. having classpath runtime bindings that change behavior). I've found that the roll-you-own intermediate solution is better (i.e. best trade-off between internal-vs-external complexity).

> The PR would be smaller, IMHO, if it basically centered around the Udmi Pod, State Handler, and the definition for what the messages look like in the queue.

If the goal is to make the PR smaller then it's easy enough to extract, but I was going for a complete PR that proved the required functionality. If if it's just a bunch of unit tests then it doesn't form a workable unit (e.g. including an example of how the interfaces are actually used somewhere).

**Review by johnrandolph — CHANGES_REQUESTED** [review]
I think this is clever but hard to follow. In my understanding of what we are doing, I think we should have a basic Message that wraps JMS. We could next have the layer that does the class/type magic. Finally, the code for the MqttPipe would be about as slim as it is, and then the localMessagePipe would be almost no code.

In the interest of moving forward I am going to approve under condition that we simplify this in the future. Maybe a good process would be to agree on the interfaces before coding them?

File: udmis/src/main/java/com/google/bos/udmi/service/messaging/MessageContinuation.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
+package com.google.bos.udmi.service.messaging;
+
+import com.google.bos.udmi.service.messaging.MessageBase.Bundle;
+import udmi.schema.Envelope;
+
+/**
+ * Represents a continuation of a received message. This allows access to ancillary message data
+ * (through the envelope), and the ability to reprocess ths message after modification.
+ */
+public class MessageContinuation {
```
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/src/main/java/com/google/bos/udmi/service/messaging/MessageContinuation.java]: Removed as discussed.

File: udmis/src/main/java/com/google/bos/udmi/service/pod/ComponentBase.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+/**
+ * Baseline functions that are useful for any other component. No real functionally, rather
+ * convenience and abstraction to keep the main component code more clear.
+ */
+public class ComponentBase {
```
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/src/main/java/com/google/bos/udmi/service/pod/ComponentBase.java]: Updated.

**Comment by grafnu** [comment]
Updated PTAL.

**Review by johnrandolph — APPROVED** [review]
